### PR TITLE
HAWQ-814. Add ShareLibraryGen utility in common library for new feature test framework

### PR DESCRIPTION
**Pull request by huor**

### CODE REVIEW

**Comment by huor** [comment]
@xunzhang, @yaoj2, @ztao1987, please review

File: src/test/feature/testlib/test_lib.cpp [hunk 1]
```diff
@@ -122,3 +123,28 @@ TEST_F(TestCommonLib, TestFileReplace) {
   util.execSQLFile("testlib/sql/template.sql",
                    "testlib/ans/template.ans");
 }
+
+TEST_F(TestCommonLib, TestShareLibraryGen) {
+  // prepare share library
+  hawq::test::ShareLibraryGen slg;
+  slg.generate("testlib", "share_library_example.c", "share_library_example.so");
```
Comment (xunzhang) [inline src/test/feature/testlib/test_lib.cpp]: I am a little confused with the interface here..

File: src/test/feature/lib/share_library_gen.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+{
+public:
+	ShareLibraryGen() = default;
+	~ShareLibraryGen() = default;
+	ShareLibraryGen(const ShareLibraryGen&) = delete;
+	ShareLibraryGen& operator=(const ShareLibraryGen&) = delete;
+
+	void generate(const std::string& src_file_dir,
+	              const std::string& src_file_name,
+	              const std::string& lib_file_name);
```
Comment (xunzhang) [inline src/test/feature/lib/share_library_gen.h]: Add doxgen comment here?

File: src/test/feature/lib/share_library_gen.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	auto src_file_name_no_extentsion = src_file_name.substr(0, src_file_name.find_last_of("."));
+	strs_src_dst["@DESTINATION_OBJ_NAME@"] = test_suite_root + src_file_name_no_extentsion + ".o";
+	strs_src_dst["@DESTINATION_LIB_NAME@"] = test_suite_root + "/" + lib_file_name;
+
+	auto make_file_tpl = feature_test_root + "/lib/Makefile.C.in";
+	auto make_file = test_suite_root + "/Makefile";
+	frep.replace(make_file_tpl, make_file, strs_src_dst);
+
+	// Generate share library
+	hawq::test::Command::getCommandStatus("cd " + test_suite_root + "; make; make clean");
```
Comment (xunzhang) [inline src/test/feature/lib/share_library_gen.cpp]: Suggestion:
`#include "lib/string_util.h"`
`auto cmd = hawq::test::stringFormat("cd %s; make; make clean", test_suite_root);`
`hawq::test::Command::getCommandStatus(cmd.c_str());`

File: src/test/feature/lib/share_library_gen.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+namespace hawq {
+namespace test {
+
+void ShareLibraryGen::generate(const string& src_file_dir,
+                               const string& src_file_name,
+                               const string& lib_file_name)
+{
+	// Generate Makefile dynamically
+	hawq::test::SQLUtility util;
+	auto feature_test_root = util.getTestRootPath();
```
Comment (xunzhang) [inline src/test/feature/lib/share_library_gen.cpp]: suggestion:
`auto hawq_home = feature_test_root + "/../../..";`
`strs_src_dst["@HAWQ_INCLUDE_DIR@"] = hawq_home + "/src/include";`
`strs_src_dst["@HAWQ_PORT_DIR@"] = hawq_home + "/src/port";`
`strs_src_dst["@POSTGRESQL_BIN@"] = hawq_home + "/src/backend/postgres";`

File: src/test/feature/testlib/test_lib.cpp (hunk 1 above)
Comment (huor) [inline src/test/feature/testlib/test_lib.cpp]: Show me your confusion:)

File: src/test/feature/lib/share_library_gen.h (hunk 2 above)
Comment (huor) [inline src/test/feature/lib/share_library_gen.h]: Cool. Added.

File: src/test/feature/lib/share_library_gen.cpp (hunk 4 above)
Comment (huor) [inline src/test/feature/lib/share_library_gen.cpp]: Sounds great.

File: src/test/feature/lib/share_library_gen.cpp (hunk 3 above)
Comment (huor) [inline src/test/feature/lib/share_library_gen.cpp]: Cool.

**Comment by xunzhang** [comment]
+1

File: src/test/feature/UDF/lib/Makefile [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+RM = rm -rf 
+
+all: $(OBJS) $(TARGET)
+
+$(OBJS): $(PROG)
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c -o $(OBJS) $(PROG)
+
+$(TARGET):
+ifeq ($(OS),Darwin)
```
Comment (xunzhang) [inline src/test/feature/UDF/lib/Makefile]: You do not need to if-else here, you should just include `src/Makefile.port`.

**Comment by xunzhang** [comment]
+1

**Comment by yaoj2** [comment]
＋1

**Comment by radarwave** [comment]
+1

**Pull request closed without merging**